module system: type merging for enum, nullOr and listOf

ericsagnes opened the pull request:
###### Motivation for this change

This adds type merging for `enum`, `nullOr` and `listOf` types.

It allow to extend type definitions for `enum`, `nullOr enum` and `listOf enum` options over multiple modules files in a decentralized manner.

Idea come from https://github.com/NixOS/nixpkgs/pull/15096#issuecomment-239925597.
#### Example use case: Input methods

This functionality can be used extend sum types defining a choice without changing the base module.

There can be only one input method enabled system-wide, so the enabled input method option is declared as an `enum` in the central input method module (`default.nix`).
Every input method has its own set of options and config and so is managed in its own module file.

Currently, options declarations are closed so the declaration in `default.nix` must contains all the possible choices:

_default.nix_

```
  enabled = mkOption {
    type    = types.nullOr (types.enum [ "ibus" "fcitx" "nabi" "uim" ]);
    default = null;
    ...
  };
```

That means that any addition or removal of input methods result in a change in the `enabled` declaration in `default.nix`.

With type merging, it is possible to make `enabled` a placeholder in the main module:

_default.nix_:

```
  enabled = mkOption {
    type    = types.nullOr (types.enum [ ]);
    default = null;
    ...
  };
```

And to extend it in other input methods modules files:

_ibus.nix_:

```
  enabled = mkOption {
    type = types.nullOr (types.enum [ "ibus" ]);
  };
```

As a result, option declaration is decentralized and modularity improved.
#### Manual

In the manual, all the files that extends or declare the type are listed in `Declared by:`, sample output:

```
i18n.inputMethod.enabled

    Select the enabled input method. Input methods is a software to input symbols that are not available on standard input devices. Input methods are specially used to input Chinese, Japanese and Korean characters.

    Type: null or one of "uim", "nabi", "ibus", "fcitx"

    Default: null

    Example: "fcitx"

    Declared by:
    <nixpkgs/nixos/modules/i18n/input-method/uim.nix>
    <nixpkgs/nixos/modules/i18n/input-method/nabi.nix>
    <nixpkgs/nixos/modules/i18n/input-method/ibus.nix>
    <nixpkgs/nixos/modules/i18n/input-method/fcitx.nix>
    <nixpkgs/nixos/modules/i18n/input-method/default.nix> 
```

---

This functionality is useful for any case where modular sum types options could make sense (display managers, window managers, desktop managers modules, ...)

cc @nbp
###### Things done
- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ericsagnes committed:
module system: extensible option types

ericsagnes commented on lib/modules.nix:
```diff
@@ -233,9 +233,7 @@ rec {
          opt.options ? description && res ? description ||
-         opt.options ? apply && res ? apply ||
-         # Accept to merge options which have identical types.
-         opt.options ? type && res ? type && opt.options.type.name != res.type.name
```
@nbp It is needed to merge `enums` because two `enum` with different parameters have different names, but I am not very sure of the implications removing this can have.
Is this safe?

nbp replied:
The goal of this check was to prevent the extension of additional option declarations with a type which is not identical.  For example, submodule option has a documentation string (`name`), thus we can extend them with additional submodules.  Another example, is with 2 identical types, such as `uniq string` can be merged with each others.

So this is technically a regression.

The generic question is more subtle than that, which is, is the merge procedure between these 2 types commutative?  But honestly, we don't want to answer this one, because in practice this does not hold true, the reason being that the order of the merge is unspecified.

With the addition of `typeMerge` function, I would recommend to drop the use of these documentation strings (`name`), in favor of easily comparable types which can be used to uniquely identify a type, and also in a way to carry over the values out-side the comparison set.

So, one potential option would be to replace the `inhabitant` attribute, by a `functor` attribute, which can be compared and merged.

```nix
{
  enum = values: mkOptionType {
    functor = { type = enum; wrapped = []; payload = [ values ]; };
  };

  loaOf = elemType: mkOptionType {
    functor = { type = loaOf; wrapped = [ elemType ]; payload = []; };
  };
}
```

Thus merging option declarations will recursively check for if the functions are identical, and if we can zip the arguments, otherwise we should throw.

What do you think?

nbp commented:
@ericsagnes , Thanks.
I will carefully review these changes on Saturday, ping me on irc if I forgot.

nbp commented:
Even with the comment above which explains why we cannot merge it as-is, this is still some nice work!

ericsagnes commented:
@nbp

Indeed, the functor idea is a great one. Thanks!
I like the new code a lot more :)

I slightly adapted your suggestion to something I found more practical to implement:

- `wrapped` and `payload` are not lists but just the `values` or `elemType` of the type to simplify the default merging
- added a binary operation to combine payloads values, `binOp`, to make the default type merging function as generic as possible

Other changes:

- the removed check is back and augmented with a condition that verifies that types are not mergeable
- I took the freedom to refactor a little the check condition to make it easier for the eyes

I tried to play with `nix-repl` but couldn't find a way to check that two functions are identical, so I restored the previous check for the moment.

```
> let a = x: 1; b = y: 1; in [ (a == a) (a == b) ]
[ false false ]
```

Maybe I am missing something obvious here?

There is always the possibility to add a string representation of the type function name to the functor in case function check is not possible, but I would prefer to have your opinion before going this way.

Note: I let the changes in a separate commit for easier review, but will squash when the PR will be ready.

ericsagnes commented:
@nbp

I added a new commit to separate type names and descriptions.
That should address the problems that `name` ambiguity had.

I also changed the `fixupOptionType` in `lib/module.nix` to make it work with the type name and functor as it was relying on the type name only.

The check removed in the first commit is now restored in its original state (beside the second commit refactor).

ericsagnes commented:
update:
- fixed a bug with the `bothHave` function.
- removed unneeded checks.

nbp commented on lib/options.nix:
```diff
@@ -92,7 +92,7 @@ rec {
           visible = opt.visible or true;
           readOnly = opt.readOnly or false;
-          type = opt.type.name or null;
+          type = opt.type.description or null;
```
Identically, update https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/installer/tools/nixos-option.sh#L259

nbp commented on lib/modules.nix:
```diff
@@ -230,12 +230,20 @@ rec {
+         bothHave "example" ||
+         bothHave "description" ||
+         bothHave "apply" ||
+         bothHave "type" && (! sameType)
```
nit:   bothHave "type" && (mergedType != null)

If they do not have the same type, then they should not be mergeable, and we should report this as an error.
Note, sameType compares the name of the top-level type, but not the wrapped type.  Thus "list" == "list", but not "list(string)" != "list(attrs)".  I do not think this would be wise to create extra strings when we can avoid, as we already have&need the functor.

nbp commented on lib/types.nix:
```diff
@@ -245,14 +286,18 @@ rec {
 
     either = t1: t2: mkOptionType {
-      name = "${t1.name} or ${t2.name}";
+      name = "either";
```
This option should have a typeMerge with 2 wrapped types.

nbp commented on lib/modules.nix:
```diff
@@ -230,12 +230,20 @@ rec {
+          mergedType = t.typeMerge t'.functor;
+          typeSet = if (bothHave "type") && sameType && (mergedType != null)
+                       then { type = mergedType; }
+                       else {};
```
If both inputs have a type, then the output should also have a type.  I suggest to remove the other parts of the condition.

nbp commented:
@ericsagnes 
> ```
> let a = x: 1; b = y: 1; in [ (a == a) (a == b) ]
[ false false ]
```
> 
> Maybe I am missing something obvious here?

Hum … Last time I tried that the result was `[ true false ]`.  Maybe this is a regression in the Nix language. (@edolstra ?)

ericsagnes commented:
@nbp 

Thanks for the review!
I added a commit that should address all the points, let me know if I missed something.

ericsagnes commented:
update: added some documentation.

The commits are kept for review, and will be squashed when the PR is ready.

@nbp Could you review the recent changes?

ericsagnes commented:
@nbp ping (sorry for spamming but I don't know who could review this PR beside you)

joachifm commented:
cc @Ericson2314 might be able to provide some feedback

Ericson2314 commented:
Hmm me? I'm not very familiar with the inner workings of the module system. From a general PL perspective, I'd consider making extensibililty opt-in--i.e. have `closedEnum` and `openEnum`--but I haven't thought about the specifics of the nixpkgs usecase.

nbp commented on lib/types.nix:
```diff
@@ -245,23 +287,46 @@ rec {
+            mt2 = t2.typeMerge t2'.functor;
+        in
+           # Eithers wrapping different types are unmergeable
+           if (t1.name != t1'.name) || (t2.name != t2'.name)
```
Testing the name in the `either` type should no longer be necessary iff this is being handler by the `defaultTypeMerge` function.  In which case, only `mt1 == null || mt2 == null` would be necessary.

nbp commented on lib/types.nix:
```diff
@@ -245,23 +287,46 @@ rec {
         check = flip elem values;
         merge = mergeOneOption;
+        typeMerge = defaultTypeMerge functor;
+        functor = { type = enum; wrapped = null; payload = values; binOp = a: b: a ++ b; };
```
nit: Should we avoid duplicates in the enumerated values?

nbp commented on lib/types.nix:
```diff
@@ -245,23 +287,46 @@ rec {
       merge = mergeOneOption;
+      typeMerge = f':
+        let t1  = head   functor.wrapped;
+            t2  = elemAt functor.wrapped 1;
```
Why should these be re-computed, as we already know them as `t1` and `t2` from the `either` function arguments?

nbp commented on lib/types.nix:
```diff
@@ -245,23 +287,46 @@ rec {
+        let t1  = head   functor.wrapped;
+            t2  = elemAt functor.wrapped 1;
+            t1' = head   f'.wrapped;
+            t2' = elemAt f'.wrapped 1;
```
We should check that `f.type.name == name`  before interpreting `f'.wrapped` as being a list.

nbp commented on lib/types.nix:
```diff
@@ -245,23 +287,46 @@ rec {
+             then null
+           else functor.type
+                  (if mt1 != null then mt1 else t1)
+                  (if mt2 != null then mt2 else t2);
```
This is not a commutative behaviour, either both types should be mergeable, or we should not be able to merge this either type.

nbp reviewed (changes requested):
Sorry for the delay.

I think these changes are going in the right direction. Still, I do not think it is yet acceptable.

ericsagnes commented:
Thanks for the review!

I addressed all the points in the latest commit, it looks cleaner now.

> We should check that f.type.name == name before interpreting f'.wrapped as being a list.

There was a problem with that as `f.type` is the unevaluated type function, so it is not easy to extract `name` from it.

I chose to add the `name` to the functor but let me know if there is a better way.

The commit also introduce a few other changes:

- `defaultFunctor` is now a function taking the type name.
- `defaultFunctor` `type` is automatically derivated from `name` (when possible).
- As `typeMerge` is now set by default, type merging was explicitly disabled for `submodule`
- explicit `typeMerge` for `separatedString` as string types with a different separator should not be merged
- more checks in `defaultTypeMerge`

nbp commented on lib/types.nix:
```diff
@@ -17,10 +17,43 @@ rec {
+       then f.type wrapped
+    # value types
+    else if (f.payload != null && f'.payload != null) && (payload != null)
+       then f.type payload
```
note: This assumes that `null` is not a valid payload, but I am fine with this restriction.

nbp commented on lib/types.nix:
```diff
@@ -52,29 +99,39 @@ rec {
+      description = "string";
       check = isString;
       merge = loc: defs: concatStringsSep sep (getValues defs);
+      functor = (defaultFunctor name) // { inherit sep; };
```
`sep` does not have any meaning as part of a `functor`.  I think what you could do is:

```nix
{
  functor = (defaultFunctor name) // {
    payload = sep;
    binOp = sepLhs: sepRhs:
       if sepLhs == sepRhs then sepLhs
       else null;
  };
}
```

nbp commented on lib/types.nix:
```diff
@@ -52,29 +99,39 @@ rec {
+      # do not type merge separated strings with different separators
+      typeMerge = f':
+        if (name == f'.name) && (sep == f'.sep)
+        then functor.type sep
```
With the new functor from above, the `defaultTypeMerge` should be enough.  So removing this code should be fine.

nbp commented on nixos/doc/manual/development/option-declarations.xml:
```diff
@@ -65,4 +65,68 @@ options = {
+      linkend='ex-option-declaration-eot-backend-bar' />, <xref 
+      linkend='ex-option-declaration-eot-backend-baz' />). As a result, 
+    <literal>services.foo.backend</literal> option values can be added or 
+    removed without changing the service module file.</para>
```
nit: we cannot remove, unless we do not include a module.

nbp commented on nixos/doc/manual/development/option-declarations.xml:
```diff
@@ -65,4 +65,68 @@ options = {
+
+  <para>As an example, we will take the case of a service 
+    <literal>foo</literal>, that has multiple backends: <literal>bar</literal> 
+    and <literal>baz</literal>.
```
foo, bar, and baz might not be easy to follow, maybe taking a concrete example might help get a better understanding of the feature provided by extensible option types.

For examples:
 * service: audio; backends: alsa, pulseaudio.
 * service: webserver; backends: apache, nginx.
 * service: init-system; backends: upstart, systemd, initd.
 * service: display manager; backends: kdm, slim.

nbp commented on nixos/doc/manual/development/option-declarations.xml:
```diff
@@ -65,4 +65,68 @@ options = {
 </para>
 
+<section xml:id="sec-option-declarations-eot"><title>Extended Option 
+    Types</title>
```
nit: I would replace `Extended` by `Extensible`, and do the same for all the references below as well.
`Extended` would apply to the result, not the property of a type.

nbp commented on lib/types.nix:
```diff
@@ -236,6 +306,8 @@ rec {
         getSubModules = opts';
         substSubModules = m: submodule m;
+        # disabling type merging for submodules
+        typeMerge = f: f': null;
```
I highly doubt this would work with the current NixOS modules, as we at least use that for the `fileSystems`.

Currently the merging process of submodules is handled by `mergeOptionDecls`, but we still expects to be able to merge the type.  The `opts` value is similar to the list of enumerated values of the `enum` type, but as the submodules are extracted, I guess you could just define the following `functor`:

```nix
{
  functor = (defaultFunctor name) // {
    # Merging of submodules is done as part of mergeOptionDecls, as we have to annotate
    # each submodule with its location.
    payload = [];
    binOp = lhs: rhs: [];
  };
}
```

nbp reviewed (changes requested):
Apart from the submodule type `typeMerge` function, which would fail on NixOS, this patch looks perfect.

This feature would be awesome once we could use it for all the examples listed in the documentation nit below.

ericsagnes commented:
Thanks again for the review!

Addressed all the points in the last commit and added a changelog.

nbp commented on nixos/doc/manual/development/option-declarations.xml:
```diff
@@ -65,4 +65,91 @@ options = {
+  <para>Extensible option types is a feature that allow to extend certain types 
+    declaration through multiple module files.
+    This feature only work with a restricted set of types, namely 
+    <literal>enum</literal> and its composed forms.</para>
```
nit: namely *enum* and *submodules* and any composed forms of them.

nbp commented on nixos/doc/manual/development/option-types.xml:
```diff
@@ -73,12 +73,26 @@
+ <section><title>Value Types</title>
+
+   <para>Value types are type that take a value parameter. The only value type 
+     in the library is <literal>enum</literal>.</para>
```
nit: *separatedString* and *submodules* are also value types.

nbp commented on nixos/doc/manual/release-notes/rl-1703.xml:
```diff
@@ -75,7 +75,9 @@ following incompatible changes:</para>
   <listitem>
-    <para></para>
+    <para>Module type system have a new extensible option types feature
+      that allow to extend certain types through multiple modules files.
```
nit: to extend certain types, such as *enum*, through multiple option declarations of the same option across multiple modules.

nbp commented on nixos/doc/manual/development/option-types.xml:
```diff
@@ -388,6 +399,53 @@ code before creating a new type.</para>
+          <term><varname>payload</varname></term>
+          <listitem><para>Holds the value parameter for value types. 
+              The only type in the library that has a 
+              <literal>payload</literal> is the <literal>enum</literal> 
```
nit: not the only type.

nbp reviewed (approved):
Apart from the documentation nits, the technical part is good and mergeable.

ericsagnes commented:
Thanks! Updated the documentation and squashed the commits so this should be ready to merge.

nbp commented:
@ericsagnes  Awesome work! Thanks for your dedication and patience :)

ericsagnes commented:
Thanks for merging!
